include/grpc++/impl/codegen/completion_queue.h

```diff
@@ -109,6 +111,26 @@ class CompletionQueue : private GrpcLibraryCodegen {
     TIMEOUT     ///< deadline was reached.
   };
 
+  /// EXPERIMENTAL
+  /// First executes \a F, then reads from the queue, blocking up to
+  /// \a deadline (or the queue's shutdown).
+  /// Both \a tag and \a ok are updated upon success (if an event is available
+  /// within the \a deadline).  A \a tag points to an arbitrary location usually
+  /// employed to uniquely identify an event.
+  ///
+  /// \param F[in] Function to execute before calling AsyncNext on this queue.
+  /// \param tag[out] Upon sucess, updated to point to the event's tag.
+  /// \param ok[out] Upon sucess, true if read a regular event, false otherwise.
+  /// \param deadline[in] How long to block in wait for an event.
+  ///
+  /// \return The type of event read.
+  template <typename T>
+  NextStatus DoThenAsyncNext(const std::function<void(void)>& F, void** tag,
```

Comment:
can we do:

```
template <typename T, typename F>
NextStatus DoThenAsyncNext(F&& f, /* ... rest of args ... */) {
  // ... rest of code
}
```

and avoid unnecessarily instantiating a std::function (since this is not for free, and lambda's pretty much are)